### PR TITLE
Fix no-extraneous-dependencies: Take 2

### LINKED ISSUE #1001: Prevent missing `@shopify/*` packages from going out to releases
## Overview

Packages found in `./packages` often import one another. For example, `@shopify/react-server-webpack-plugin` makes use of `@shopify/ast-utilities`. In these cases, the imported package must be listed in the importing package's `package.json`. This can be easy to forget as there are no build, development, or publishing error reported until the package is released and installed in a consuming project. The consuming project will get an error, such as `'@shopify/ast-utilities' modules not found`.

An easy temporary fix when this happens is to install the missing package in the consuming project, but the permanent fix is to add ’@shopify/ast-utilities' to `@shopify/react-server-webpack-plugin` package.json#dependencies. This will require a re-release, which can be time-consuming.

### 1. Lint

We can lint for missing packages and fail the lint step. A rule exists called `import/no-extraneous-packages` from `eslint-plugin-import`, but this rule is recognizing the @shopify namespace imports as “internal” (not “external”) causing them to be skipped over/ not reported as extraneous.

#### To see the problem:

- Hack the typescript config in node modules and turn on the rule. (I am not sure why we have it turned off at the moment, the comment says it has something to do with not supporting typescript, but I haven't experienced any issues with the rule https://github.com/Shopify/eslint-plugin-shopify/blob/master/lib/config/typescript.js#L41) 
- Remove the `@shopify/ast-utilities` package from `react-server-webpack-plugin` and `run yarn lint '*/**/react-server-webpack-plugin/**/**.ts’` and you should see 2 errors. This is wrong because there is no indication of the missing `ast-utilities` package we just deleted.
- Now run ```$ node --inspect-brk node_modules/eslint/bin/eslint.js '*/**/react-server-webpack-plugin/**/**.ts’``` from the quilt repo.
- Go to `chrome://inspect`  and adding a breakpoint In on line ~123 of the no-extraneous-import rule. 
- Change the import type condition to check for both external and internal, you should now see 3 errors instead of 2.

I haven't been able to figure out why this is happening. It could be a configuration problem (I tried a number of different settings) or could require a patch to the repo.

### 2. Run a script/test

This can be wrapped into a jest test, but something that runs in CI and checks to make sure that any imported packages are also listed in the package.json. 

A rough example below.

```tsx
const path = require('path');
const {readdirSync} = require('fs-extra');
const grep = require('simple-grep');
const uniq = require('lodash/uniq');

const packagesPath = path.resolve(__dirname, '..', 'packages');
const packageNames = readdirSync(packagesPath).filter(
  file => !file.includes('.'),
);

packageNames.forEach(checkDependencies);

function checkDependencies(package) {
  const packageJson = path.resolve(packagesPath, package, 'package.json');
  let pkg;

  try {
    pkg = require(packageJson);
  } catch (error) {
    // swallow error
  }

  if (!pkg) {
    return;
  }

  const {dependencies = {}, devDependencies = {}, peerDependencies = {}} = pkg;
  const allDependencies = [
    ...Object.keys(dependencies),
    ...Object.keys(devDependencies),
    ...Object.keys(peerDependencies),
  ];
  const missingDependencies = [];

  grep(`@shopify/`, path.resolve(packagesPath, package), function(list) {
    list.forEach(item => {
      if (path.extname(item.file) !== '.ts') {
        return;
      }

      item.results.forEach(({line}) => {
        const parts = line.split('from');

        if (parts.length !== 2) {
          return;
        }

        const imported = parts[parts.length - 1]
          .replace(/'/g, '')
          .replace(/;/, '')
          .trim();

        if (!allDependencies.includes(imported)) {
          missingDependencies.push(imported);
        }
      });
    });

    if (missingDependencies.length > 1) {
      console.warn(
        `Missing Dependencies found in ${package}. You must include ${uniq(
          missingDependencies,
        ).join(', ')} in the package.json for this package.`,
      );
    }
  });
}
```

### PR DESCRIPTION
**Pull request by cartogram**
## Description

Fixes #1001
Follow up from https://github.com/Shopify/quilt/pull/1026

This PR is split into [2 commits](https://github.com/Shopify/quilt/pull/1064/commits). The [first](https://github.com/Shopify/quilt/pull/1064/commits/11345db3fccbe75a9c64f34d6087c5064071e327) contains only the changes required to make the `import/no-extraneous-dependencies` rule work for this repo, the [second](https://github.com/Shopify/quilt/pull/1064/commits/af001140efeb3953c4018fd0ecb4335a066792fa) is all the lint fixes required to pass CI.

The fix here was to use the `import/resolver` setting to override the typescript resolver from `plugin:shopify/typescript`. The typescript resolver was problematic because it resolves modules to their source location, causing the rule think of them as "internal" (not extraneous) modules. Creating a new configuration and extending it before extending the `plugin:shopify/typescript` tells the rule to use the node resolver, which looks for these packages in `node_modules` instead and properly sees them as "external" to the eslint rule.

The downside to having this rule enabled is that it (correctly) enforces all our `@shopify/*` dependencies to be listed when they are used and thus increases the dependency web and number of "patch" dependency bumps needed (releasing a new version of dependency `x` causes all related dependencies that use `x`  to need a new version). The largest offender of this was `@shopify/react-testing`, which is used in all of the `@shopify/react-*` dependencies. 

The fix for this was to add an override for test files that disables this rule. We don't care if `@shopify/react-testing` is listed or not since its really just needed for development and will never cause a missing dependency error when installed in a consuming package, which is the only thing we _really_ care about here.

### CODE REVIEW

File: packages/graphql-persisted/package.json [hunk 1]
```diff
@@ -29,7 +29,8 @@
     "koa-compose": ">=3.0.0 <4.0.0"
   },
   "devDependencies": {
-    "@shopify/sewing-kit-koa": "^6.1.2"
+    "@shopify/sewing-kit-koa": "^6.1.2",
+    "@types/koa": "^2.0.45"
```
Comment (cartogram) [inline packages/graphql-persisted/package.json]: This library, `@shopify/graphql-persisted`, uses the `Context` from `koa` and would require the consumer to have `koa` installed in their project. There were a number of packages that met this criteria, and I use the following method to resolve them: 1) Add the packages the host is required to have in the `peerDependencies` with a pretty loose version range. and 2) add the types packages used to the `devDependencies`.

Why? I added the `peerDependency` because it is needed yet unused in this library. I added a loose version range to reduce the chances of 2 version of the package been installed as a result of this library.

I added the types in `devDependencies` because they should be there in development, but are not exported in this library. However, there is a middleware `persistedOperationAssociationMiddleware` that uses the `Context` type in one of its arguments and _that_ function _is_ being exported. For this reason, I wonder if the types should be moved to dependencies. If anyone can clarify this, please comment.

File: .eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      files: [
+        '**/*/test/**/*.ts',
+        '**/*/test/**/*.tsx',
+        '**/*/tests/*.ts',
+        '**/*/tests/*.tsx',
+        '*.test.ts',
+        '*.test.tsx',
+      ],
+      rules: {
+        'shopify/jsx-no-hardcoded-content': 'off',
```
Comment (cartogram) [inline .eslintrc.js]: Here we turn this rule off to prevent requiring packages such as `@shopify/react-testing` in ever react-based package.

File: packages/jest-koa-mocks/package.json [hunk 3]
```diff
@@ -23,13 +23,13 @@
   },
   "homepage": "https://github.com/Shopify/quilt/blob/master/packages/jest-koa-mocks/README.md",
   "dependencies": {
-    "koa": "^2.5.0",
+    "koa": "^2.8.2",
```
Comment (cartogram) [inline packages/jest-koa-mocks/package.json]: Bringing our `koa` and `@types/koa` to be more consistent across packages in order to remove duplicate versions.

File: packages/graphql-persisted/package.json [hunk 4]
```diff
@@ -24,6 +24,7 @@
   "dependencies": {
     "@types/koa-bodyparser": "*",
     "@types/koa-compose": "*",
+    "@types/koa": "^2.0.50",
```
Comment (cartogram) [inline packages/graphql-persisted/package.json]: This library uses the `Context` from `koa` and would require the consumer to have `koa` installed in their project. There were a number of packages that met this criteria, and I use the following method to resolve them: 1) Add the packages the host is required to have in the `peerDependencies` with a pretty loose version range, and 2) add the types packages used to the dependencies.

Why? I added the peerDependency because it is needed (yet unused) in this library. You can't really use a `koa` middleware as intended without `koa`, but you also don't need `koa` to write them. They are, more or less, just functions. I added a loose version range to reduce the chance 2 versions of the package will be installed, in this case `koa`, as a result of this library.

I added the types in dependencies because there is a middleware, `persistedOperationAssociationMiddleware`, that uses the `Context` type in one of its arguments and that function is being exported. Even though the use of the type is tangental, we are not exporting the type itself or expecting consumers will use it, they will use the function that uses the type. For this reason, it needed to be in dependencies.

File: packages/react-html/package.json [hunk 5]
```diff
@@ -29,7 +29,8 @@
     "@shopify/useful-types": "^2.0.1",
     "@types/multistream": "^2.1.1",
     "multistream": "^2.1.1",
-    "tslib": "^1.9.3"
+    "tslib": "^1.9.3",
+    "serialize-javascript": "^2.1.0"
```
Comment (cartogram) [inline packages/react-html/package.json]: This was a strange one to be missing.

File: packages/graphql-persisted/package.json (hunk 4 above)
Comment (lemonmade) [reply packages/graphql-persisted/package.json]: Good rationale. Is this version range too restrictive, though?

File: packages/react-i18n/src/i18n.ts [hunk 6]
```diff
@@ -1,3 +1,6 @@
+import {memoize as memoizeFn} from '@shopify/function-enhancers';
+import {memoize} from '@shopify/decorators';
+import {languageFromLocale, regionFromLocale} from '@shopify/i18n';
```
Comment (cartogram) [inline packages/react-i18n/src/i18n.ts]: With the resolver changes and re-classification of these libraries (from "internal" to "external"), they are being rearranged as a result of `import/order`.

File: packages/react-router/src/components/Link/Link.tsx [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import * as React from 'react';
+import React from 'react';
```
Comment (cartogram) [inline packages/react-router/src/components/Link/Link.tsx]: This is unrelated to the purpose of this PR, just cleaning up while in here.

File: packages/react-server/src/logger/logger.ts [hunk 8]
```diff
@@ -3,7 +3,10 @@ import {Context, Request} from 'koa';
 import chalk from 'chalk';
 import {KoaNextFunction} from '../types';
 
-export const LOGGER = Symbol('logger');
+// Have to cast as `any` or typescript will complain
+// about using a "unique symbol as an index type".
+// https://github.com/microsoft/TypeScript/issues/1863
+export const LOGGER: string = Symbol('logger') as any;
```
Comment (cartogram) [inline packages/react-server/src/logger/logger.ts]: Open to other ways around this. It is a pattern we use a lot so unfortunate that we have to cast.

File: packages/koa-shopify-auth/package.json [hunk 9]
```diff
@@ -32,10 +32,10 @@
   "devDependencies": {
     "@shopify/jest-dom-mocks": "^2.8.0",
     "@shopify/jest-koa-mocks": "^2.1.5",
-    "@types/koa": "^2.0.44",
+    "@types/koa": "^2.0.50",
     "@types/koa-compose": "*",
     "@types/safe-compare": "^1.1.0",
-    "koa": "^2.5.0"
+    "koa": "^2.8.2"
```
Comment (lemonmade) [inline packages/koa-shopify-auth/package.json]: Just a heads up that you don't generally need to update dependencies in this way, using `yarn-deduplicate` can get you down to a single set of installed dependencies without forcing more restrictive version ranges

File: packages/react-cookie/src/CookieUniversalProvider.tsx [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import React from 'react';
-import {useNetworkManager} from '@shopify/react-network';
 import {useLazyRef} from '@shopify/react-hooks';
+import {useNetworkManager} from '@shopify/react-network';
```
Comment (lemonmade) [inline packages/react-cookie/src/CookieUniversalProvider.tsx]: Why does this kind of re-order need to happen? Aren't they both external?

File: packages/react-html/package.json (hunk 5 above)
Comment (lemonmade) [reply packages/react-html/package.json]: Why?

File: packages/react-i18n/package.json [hunk 11]
```diff
@@ -34,6 +34,8 @@
     "intl-pluralrules": "^0.2.1"
   },
   "dependencies": {
+    "@babel/template": "^7.6.0",
+    "@babel/traverse": "^7.6.2",
```
Comment (lemonmade) [inline packages/react-i18n/package.json]: Hmm, I'm not sure if these are strictly dependencies. Aren't they just used for types? Babel should generally be a peer/ optional dependency IMO

File: packages/sewing-kit-koa/src/middleware.ts [hunk 12]
```diff
@@ -17,7 +17,10 @@ export interface Options {
   manifestPath?: string;
 }
 
-const ASSETS = Symbol('assets');
+// Have to cast as `any` or typescript will complain
+// about using a "unique symbol as an index type".
+// https://github.com/microsoft/TypeScript/issues/1863
```
Comment (lemonmade) [inline packages/sewing-kit-koa/src/middleware.ts]: I don't understand, what changed that caused this?

File: packages/react-html/package.json (hunk 5 above)
Comment (cartogram) [reply packages/react-html/package.json]: This library uses it heavily so I'm surprised the library works without it listed. I think its because `react-serialize` has it.

File: packages/sewing-kit-koa/src/middleware.ts (hunk 12 above)
Comment (cartogram) [reply packages/sewing-kit-koa/src/middleware.ts]: @lemonmade the change from `types/koa` from `2.0.49 => 2.0.50`. My guess is this change https://github.com/DefinitelyTyped/DefinitelyTyped/commit/80c6d507792b9706617ca27e4fa5ec9e314db8a5#diff-7605d695f1da86ef8f6a929780113155R724

Comment (cartogram) [reply packages/sewing-kit-koa/src/middleware.ts]: `State` is no longer `any`

Comment (lemonmade) [reply packages/sewing-kit-koa/src/middleware.ts]: Can we cast `ctx.state` as `any` then? I don't like a type that is a straight up lie when we could switch to looser typing in one spot instead.

File: packages/react-cookie/src/CookieUniversalProvider.tsx (hunk 10 above)
Comment (cartogram) [reply packages/react-cookie/src/CookieUniversalProvider.tsx]: Yeah good question, this looks like a mistake.

File: packages/jest-koa-mocks/src/create-mock-context/create-mock-context.ts [hunk 13]
```diff
@@ -109,6 +109,6 @@ export default function createContext<
 
   // ctx.request.body is a common enough custom property for middleware to add that it's handy to just support it by default
   context.request.body = requestBody;
-  context.request.rawBody = rawBody;
+  context.request.rawBody = rawBody || '';
```
Comment (jgodson) [inline packages/jest-koa-mocks/src/create-mock-context/create-mock-context.ts]: could we set this as a default argument on line 53 instead? (like `url` is)

**Comment by BPScott** [comment]
I wonder if there's some upstream PR we can make to https://github.com/alexgorbatchev/eslint-import-resolver-typescript to allow support for our monorepo use case

File: .eslintrc.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+          },
+        ],
+      },
+    },
+    {
+      files: [
+        '**/*/test/**/*.ts',
+        '**/*/test/**/*.tsx',
+        '**/*/tests/*.ts',
+        '**/*/tests/*.tsx',
```
[the remark this answers is not in the document]
Comment (cartogram) [reply .eslintrc.js]: Moving 2 files (example: `react-i18n/src/e2e/translations.test.tsx` => `react-i18n/src/test/e2e/translations.test.tsx`) gets us just the 4 patterns. The test/tests thing can be resolved in a follow-up.

File: config/eslint.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+// We need to specify the import/resolver setting in an extendable
```
Comment (BPScott) [inline config/eslint.js]: I'm surprised this works - rules later in the config overrode earlier ones - it's how you can enable a rule in a config, then disable it in your local eslint config. I would have thought settings would work in the same way.

Comment (lemonmade) [reply config/eslint.js]: I think the import plugin iterates over the setting object, which gets merged together by ESLint's config. However, it merges in the order of plugin addition, and then iterates over the object (iteration happens in insertion order), so resolvers added by earlier plugins are actually handled first.

File: .eslintrc.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    {
+      files: [
+        '**/test/**/*.ts',
+        '**/test/**/*.tsx',
+        '**/tests/**/*.ts',
+        '**/tests/**/*.tsx',
+      ],
+      rules: {
+        'shopify/jsx-no-hardcoded-content': 'off',
+        'import/no-extraneous-dependencies': 'off',
```
Comment (lemonmade) [inline .eslintrc.js]: Can we leave a note for why we turn this off (we should do this always but alas)

File: packages/address/src/tests/AddressFormatter.test.ts [hunk 17]
```diff
@@ -1,5 +1,5 @@
-import {fetch} from '@shopify/jest-dom-mocks';
 import {Address} from '@shopify/address-consts';
+import {fetch} from '@shopify/jest-dom-mocks';
```
Comment (lemonmade) [inline packages/address/src/tests/AddressFormatter.test.ts]: Did you explain why this would need to go below `@shopify/address-consts`?

File: packages/graphql-persisted/package.json [hunk 18]
```diff
@@ -23,10 +23,11 @@
   "homepage": "https://github.com/Shopify/quilt/blob/master/packages/graphql-persisted/README.md",
   "dependencies": {
     "@types/koa-bodyparser": "*",
-    "@types/koa-compose": "*",
+    "@types/koa-compose": "^3.2.2",
+    "@types/koa": "^2.0.0",
     "apollo-link": ">=1.0.0 <2.0.0",
     "koa-bodyparser": ">=4.0.0 <5.0.0",
-    "koa-compose": ">=3.0.0 <4.0.0"
+    "koa-compose": "^4.1.0"
```
Comment (lemonmade) [inline packages/graphql-persisted/package.json]: Do we need to restrict the version range on this?

File: packages/graphql-persisted/package.json [hunk 19]
```diff
@@ -23,10 +23,11 @@
   "homepage": "https://github.com/Shopify/quilt/blob/master/packages/graphql-persisted/README.md",
   "dependencies": {
     "@types/koa-bodyparser": "*",
-    "@types/koa-compose": "*",
+    "@types/koa-compose": "^3.2.2",
```
Comment (lemonmade) [inline packages/graphql-persisted/package.json]: Why did this change?

File: packages/jest-dom-mocks/package.json [hunk 20]
```diff
@@ -30,7 +30,8 @@
     "fetch-mock": "^6.3.0",
     "lolex": "^2.7.5",
     "promise": "^8.0.3",
-    "tslib": "^1.9.3"
+    "tslib": "^1.9.3",
+    "lodash": "^4.17.15"
```
Comment (lemonmade) [inline packages/jest-dom-mocks/package.json]: Can this be looser?

File: packages/koa-shopify-auth/package.json [hunk 21]
```diff
@@ -24,16 +24,16 @@
   "homepage": "https://github.com/Shopify/quilt/blob/master/packages/koa-shopify-auth/README.md",
   "dependencies": {
     "@shopify/network": "^1.4.2",
-    "koa-compose": ">=3.0.0 <4.0.0",
+    "koa-compose": "^4.1.0",
```
Comment (lemonmade) [inline packages/koa-shopify-auth/package.json]: Same question about these dependencies

File: packages/react-app-bridge-universal-provider/package.json [hunk 22]
```diff
@@ -32,7 +32,6 @@
   },
   "devDependencies": {
     "@shopify/react-effect": "^3.2.3",
```
Comment (lemonmade) [inline packages/react-app-bridge-universal-provider/package.json]: 🔥 this one too?

File: packages/react-effect-apollo/package.json [hunk 23]
```diff
@@ -24,7 +24,8 @@
   },
   "homepage": "https://github.com/Shopify/quilt/blob/master/packages/react-effect-apollo/README.md",
   "dependencies": {
-    "@shopify/react-effect": "^3.2.3"
+    "@shopify/react-effect": "^3.2.3",
+    "prop-types": "^15.7.2"
```
Comment (lemonmade) [inline packages/react-effect-apollo/package.json]: Should be able to delete all of these because I deprecated the package

File: packages/react-server/package.json [hunk 24]
```diff
@@ -33,14 +33,18 @@
...
     "@shopify/react-network": "^3.3.1",
     "@shopify/sewing-kit-koa": "^6.1.2",
+    "@shopify/useful-types": "^1.3.0",
     "chalk": "^2.4.2",
     "koa": "^2.5.0",
     "koa-compose": "^4.1.0",
-    "koa-mount": "^4.0.0"
+    "koa-mount": "^4.0.0",
+    "fs-extra": "^8.1.0",
+    "isomorphic-fetch": "2.2.1"
```
Comment (lemonmade) [inline packages/react-server/package.json]: Is this actually used for it directly?

File: packages/react-server/package.json [hunk 25]
```diff
@@ -33,14 +33,18 @@
     "@shopify/react-hydrate": "^1.1.4",
     "@shopify/react-network": "^3.3.1",
     "@shopify/sewing-kit-koa": "^6.1.2",
+    "@shopify/useful-types": "^1.3.0",
     "chalk": "^2.4.2",
     "koa": "^2.5.0",
     "koa-compose": "^4.1.0",
-    "koa-mount": "^4.0.0"
+    "koa-mount": "^4.0.0",
+    "fs-extra": "^8.1.0",
```
Comment (lemonmade) [inline packages/react-server/package.json]: What is this used for?

Comment (cartogram) [reply packages/react-server/package.json]: It is used for checking if there is a Gemfile, when getting the manifest file.

```

function getManifestPath(root: string) {
  const gemFileExists = pathExistsSync(join(root, 'Gemfile'));
  if (!gemFileExists) {
    return;
  }

  // eslint-disable-next-line no-process-env
  return process.env.NODE_ENV === 'development'
    ? `tmp/sewing-kit/sewing-kit-manifest.json`
    : `public/bundles/sewing-kit-manifest.json`;
}
```

File: packages/react-server/package.json (hunk 24 above)
Comment (cartogram) [reply packages/react-server/package.json]: It isn't no.

Comment (cartogram) [reply packages/react-server/package.json]: Might make more sense as a peer dependency.

File: packages/react-server/package.json (hunk 25 above)
Comment (BPScott) [reply packages/react-server/package.json]: Looks like it's an alias for the built in fs.existsSync.

https://github.com/jprichardson/node-fs-extra/blob/master/lib/path-exists/index.js#L11

File: packages/koa-performance/package.json [hunk 26]
```diff
@@ -27,14 +27,15 @@
     "@shopify/network": "^1.4.2",
     "@shopify/performance": "^1.2.0",
     "@shopify/statsd": "^1.1.0",
+		"@types/koa": "^2.0.0",
```
Comment (lemonmade) [inline packages/koa-performance/package.json]: Weird spacing
